design/velero-client-download-apiserver_design.md

```diff
@@ -0,0 +1,128 @@
+This assumes the target storing the data is always accessible from the client side. This is not always true, especially in the on premise environments.
+Velero may also save the data in non object stores in the future so an object store may not be available for the CLI to download from.
+
+This design proposes to add a new endpoint on the Velero server that CLI users can connect to and download data from without having to connect to an object store directly.
+
+## Background
+Velero CLI commands such as following download data from the object store directly today.
+```
+velero backup download
+velero describe
+velero backup logs
+```
+We want commands to work even when the object store is not directly accessible from the client side. Examples include in-cluster minio, file system store, nfs server.
+Even if the object store is accessible from the client side, today things like cacert and insecure-skip-tls-verify flags are specified manually from the client even though the Velero server already has the information.
+
+This design allows velero CLI to request files from any object store plugin that velero server can call GetObject() without creating a signedUrl first.
+## Goals
+- Enable Velero CLI users to download data from the Velero server without having to create signedUrl for scenarios where the storage provider is not able to create signedUrl accessible from the velero CLI client side.
+
+## Non Goals
+- Making the download server the default way to download data from object store.
+
+## High-Level Design
+Create an endpoint on the Velero server that CLI users can connect to and download data from without requiring signed URL from object store directly when configured explicitly or as a fallback.
+
+### Velero Download Server URL generation using Service to Expose Download Server
+
+Upon enabling Download Server API, Velero will create a service of type that user specifies, which will expose the download server to the outside world. The download server will proxy requested data from the object store and stream it back to the client using one of the available service types (ClusterIP/NodePort/LoadBalancer/ExternalName). User can also precreate a service and pass the service name to velero during install time with `velero install --veleroServerServiceName=veleroDownloadServerServiceName` to use the precreated service to expose the download server.
+
+If the service can also be reached via a fully qualified domain name (FQDN), user can pass the FQDN to velero during install time with `velero install --veleroServerFQDN=dnsNameToVeleroServerService.com`. Velero Download Server will use FQDN in-place of values auto discovered from created service (ie. NodePort/Cluster IP)
+
+This approach may also [work on KinD clusters](https://kind.sigs.k8s.io/docs/user/loadbalancer/)
+
+The current DownloadRequest CR status `veleroServerURL` field will be populated to use the Download Server URL instead of the signed object store URL where available.
```

Comment:
> IIUC @ywk253100, you are advocating that for some object stores, we would want to use the signed URL that is returned and not the download request when it is not needed?

No, What I'm saying is that we always read the URL from the `downloadURL` field of the download request rather than adding a new `veleroServerURL`.

> If that is the case I agree, I wonder how we will determine that though?

We can add a new field in the BSL spec to determine how to get the URL.

Reply:
I like this idea.

Reply:
> decides the correct way to download the data

Not sure if there's an easy way to determine from DR reconciler what cli has access to.. CLI can perhaps create another CR saying prior attempt was unsuccessful so the next DR url returned would be download server url.

Reply:
If we add a new field (e.g. `downloadViaVeleroServer` bool) to the BSL spec, the DR reconciler can determine which URL (the URL of direct access or via the Velero server) should be returned. 
And the CLI doesn't need to care about this, it downloads the data from the `downloadURL` of DR.